src/unitxt/operators.py

```diff
@@ -213,6 +213,109 @@ def process(
+    WholeInputFormatter makes this (verbalization of the) input into a whole-input-to-the-model, by combining it
+    with instruction and demos, read from the input instance, through both formatting args
+    formatting arguments.
+    Following the Renderers action, the formatted model input is fed into field "source" (overwriting the field)
+    Also following renderers, this operator pops fields "instruction" and "demos" from the input instance.
+
+    Args:
+        demos_field (str): the name of the field that contains the demos, being a list of dicts, each with "source" and "target" keys
+        demo_format (str): formatting string for a single demo, combining fields source and target
+        model_input_format (str) overall output format, combining instruction, demos, and source
+
+    Example:
+        when input instance:
+        {
+            "source": "1+1",
+            "target": "2",
+            "instruction": "solve the math exercises",
+            "demos": [{"source": "1+2", "target": "3"}, {"source": "4-2", "target": "2"}]
+        }
+        is process-ed by
+        model_input_formatter = ModelInputFormatter(
+            demos_field="demos",
+            demo_format="User: {source}\nAgent: {target}\n\n",
+            model_input_format="{demos}User: {instruction}{source}\nAgent: ",
+        )
+        the resulting instance is:
+        {
+            "target": "2",
+            "source": "User: 1+2\nAgent: 3\n\nUser: 4-2\nAgent: 2\n\nUser: solve the math exercises\n\n1+1\nAgent: ",
+        }
+    """
+
+    demos_field: str = "demos"
+    demo_format: str = (
+        "{source}\n{target}\n\n"  #  example: "User: {source}\nAgent: {target}\n\n"
+    )
+    model_input_format: str = "{instruction}{demos}{source}\n"
+
+    @staticmethod
+    def _retrieve_field_and_assert_not_none(instance, field_name) -> str:
+        if field_name is not None and field_name in instance:
+            field_value = instance[field_name]
+            assert (
+                field_value is not None
+            ), f"Value in field '{field_name}' should not be none. Received instance: {instance}"
+            return field_value
+        return ""
+
+    def process(
+        self, instance: Dict[str, Any], stream_name: Optional[str] = None
+    ) -> Dict[str, Any]:
+        assert (
+            "source" in instance
+        ), f"field 'source' is expected to be in the input instance. Received instance: {instance}"
+        source = self._retrieve_field_and_assert_not_none(
+            instance=instance, field_name="source"
+        )
+
+        instruction = self._retrieve_field_and_assert_not_none(
+            instance=instance, field_name="instruction"
+        )
+        # pop from instance, as ICLFormat does
+        if "instruction" in instance:
+            instance.pop("instruction")
+
+        if instruction != "":
+            instruction = instruction + "\n\n"
```

Comment:
Can we remove this if? Having this type of formatting choices, should be avoided i think. If the instruction is to be followed by two new lines, then i think whoever created should be responsible to add it. 

This is regardless of my general comment about instruction coming from the instance. BTW, do you know which part of the code places the `instruction` field on the instance?

Reply:
I agree. No implicit formatting should be done.

Reply:
No problem. Now that we owe nothing to `ICLFormat`, we can do anything we want (please see my reply [here](https://github.com/IBM/unitxt/pull/423#discussion_r1437459745))